include/ginkgo/core/base/index_set.hpp

```diff
@@ -0,0 +1,444 @@
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_BASE_INDEX_SET_HPP_
+#define GKO_PUBLIC_CORE_BASE_INDEX_SET_HPP_
+
+
+#include <algorithm>
+#include <mutex>
+#include <vector>
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/base/utils.hpp>
+
+
+namespace gko {
+
+
+/**
+ * An index set class represents an ordered set of intervals. The index set
+ * contains subsets which store the starting and end points of a range,
+ * [a,b), storing the first index and one past the last index. As the
+ * index set only stores the end-points of ranges, it can be quite efficient in
+ * terms of storage.
+ *
+ * This class is particularly useful in storing continuous ranges. For example,
+ * consider the index set (1, 2, 3, 4, 5, 6, 7, 8, 10, 11, 12, 18, 19, 20, 21,
+ * 42). Instead of storing the entire array of indices, one can store intervals
+ * ([1,9), [10,13), [18,22), [42,43)), thereby only using half the storage.
+ *
+ * We store three arrays, one (subsets_begin) with the starting indices of the
+ * subsets in the index set, another (subsets_end) storing one
+ * index beyond the end indices of the subsets and the last
+ * (superset_cumulative_indices) storing the cumulative number of indices in the
+ * subsequent subsets with an initial zero which speeds up the
+ * querying. Additionally, the arrays conataining the range boundaries
+ * (subsets_begin, subsets_end) are stored in a sorted fashion.
+ *
+ * Therefore the storage would look as follows
+ *
+ * > index_set = (1, 2, 3, 4, 5, 6, 7, 8, 10, 11, 12, 18, 19, 20, 21, 42)
+ * > subsets_begin = {1, 10, 18, 42}
+ * > subsets_end = {9, 13, 22, 43}
+ * > superset_cumulative_indices = {0, 8, 11, 15, 16}
+ *
+ * @tparam index_type  type of the indices being stored in the index set.
+ *
+ * @ingroup IndexSet
+ */
+template <typename IndexType = int32>
+class IndexSet {
```

Comment:
Now that I think of it, you could probably use PolymorphicObject to provide the executor functionality and avoid the need for all of these operator overloads :)

Reply:
That is a good point!